### ISSUE TITLE
S3 Sources: Specifying the objects to ingest

### ISSUE THREAD

**Issue by quodlibetor** [MaterializeInc/materialize#5502]
An S3 bucket behaves a lot like a filesystem, and users use it a lot like one. This means that individual S3 buckets often have a large number of objects in them. When pointing materialized at an S3 bucket, users have expressed an interest in being able to specify _which_ objects should be ingested. There are several seemingly unrelated ingest schemes that have come up:

* Ingest all the objects that exist "right now", perhaps filtered to a prefix or pattern
* Read all new objects that are PUT into the S3 bucket via [the SQS/SNS notification system](https://docs.aws.amazon.com/AmazonS3/latest/dev/NotificationHowTo.html)
* Ingest objects that are present in a manifest (recalling that in a cloud environment providing a manifest file isn't really an option):
  * AWS documents [one example of this pattern for redshift](https://docs.aws.amazon.com/redshift/latest/dg/loading-data-files-using-manifest.html)
  * Some users have complex requirements that seem like they would be best solved by providing either a `TABLE` input source or the full expressivity of SQL:
    * one user has [written their own DSL to select the objects to ingest](https://github.com/MaterializeInc/materialize/issues/3773#issuecomment-754813302)
    * Another user [has their catalog in hive](https://github.com/MaterializeInc/materialize/issues/4914#issuecomment-769482520), so selecting the objects to ingest is dynamic based on an external system

Fundamentally, the problem that needs to be solved is "how do we create a set of object names to download". There are basically two options:

* Put object keys to download into a dataflow with object keys as a name and treat the S3 source as a sink for that dataflow, reading objects from it as they are inserted.
* Artisanal, hand-crafted Rust/rusoto code that selects the objects to download, with custom code paths for things like SQS notifications and manifest files. The fact that folks want to dynamically adjust the list of objects to download (as mentioned in the "complex requirements" link above) suggests that long-term we will want to support tables as a method for creating the catalog anyway.

I've created [a diagram describing these alternatives](https://app.mural.co/t/materialize7238/m/materialize7238/1611844431315/963e8dfc765bc30d4d95d6bd091b75c8295fb5ae).

The tradeoffs as I see them are that:

* The artisanal code will be faster to implement for each individual system
* The dataflow-first implementation has upfront cost and experimentation, but essentially immediately after the first implementation unlocks all the possible use-cases.

## Open Questions

* Do we need to have strong transactionality around object discovery and ingest. Currently S3 sources don't support any BYO timestamping, but we will probably want to upgrade them to support it at some point. 

  In particular, the fact that timestamps are based on wall-clock time and are not syncronized between disparate sources means that it is possible for the key-source to mark a key as available for download at the same (bit I don't believe previous to) logical time as the first rows from that object start appearing in dataflows.

  Question/discussion: https://github.com/MaterializeInc/materialize/issues/5502#issuecomment-770895995
   * Current Answer: **no**, we are fine with wall-clock time and no transactionality, with eventual upgrade to BYO.
* How will deletions work? S3 sources do not currently (and there is no short-term plan for them to) support deletion. I have [a comment describing how we might do it longer-term](https://github.com/MaterializeInc/materialize/issues/5502#issuecomment-770948130), though.
   * Current Answer: **deletions aren't planned for near-term implementation**, and it will have the standard issues that we have with deletions of append-only data sources (requiring materialization of all the source data, or an arbitrarily large amount of extra work to make precise tracing of content provenance/ID)

**Comment by quodlibetor** [comment]
Initial spec work for was documented in #4914 under the "Feature: ingesting new objects after initial import" section.

**Comment by awang** [comment]
As an alternative to hand crafted rust code, could we use UDFs or callbacks as a way to specify whether to download an object or not?

**Comment by quodlibetor** [comment]
In this case hand-crafted means "hand crafted by an integrations team member", using dataflows as the core means of specifying object keys would mean that the UDF work that is ongoing would just plug-in.

**Comment by quodlibetor** [comment]
One other thing worth pointing out is going to a dataflow-first model would potentially allow us to move even more of the existing S3 logic into dataflow, currently we have a pure-rust function that is responsible for scanning the S3 bucket, downloading, and inserting into dataflow.

We could instead split that function up into:
* download list of keys
* insert keys into dataflow
* read keys from dataflow

This true unification would allow writing arbitrary SQL to filter the objects to download, which is definitely the most flexible option on the table. I'll try and diagram this up later.

**Comment by quodlibetor** [comment]
Another way of saying my last comment is "it would be nice to have an S3 keyscan source directly" so that we could do something like:

```sql
> CREATE MATERIALIZED SOURCE s3_keys FROM S3 KEYS 'bucket';
> CREATE SOURCE s3_contents FROM S3 OBJECTS FROM VIEW s3_keys WHERE ...`
```

**Comment by umanwizard** [comment]
How does timestamping work in this model? Would `s3_contents` be expected to have the same timestamps as the underlying `s3_keys` ?

Also, would we need the `s3_contents` source to remember everything it has ever seen? So that if, for example, a key is retracted from `s3_keys`, we could retract everything we downloaded for that key...

**Comment by quodlibetor** [comment]
> Would `s3_contents` be expected to have the same timestamps as the underlying `s3_keys`?

This is a good question, I hadn't considered that we even potentially *could* do that, but it is technically possible.

I don't think we should try to make timestamps align for two reasons:

* Keys will need to be available in the key-source dataflows in order for the content dataflow to be able to use them. I believe that that means that the key-source dataflow *must* drop its capabilities for the keys to show up in the key-sink of the content-source.
* Additionally, the amount of wall-clock time that it will take to download objects from buckets is potentially unbounded -- even aside from the fact that S3 objects [can be up to 5 TB](https://aws.amazon.com/blogs/aws/amazon-s3-object-size-limit/), there is the risk of network partitions, etc. That could block *all* dataflows.

I do think it might be neat to allow something like that, but it probably depends on multidimensional timestamps or something else that is a bit further down the line.

> Also, would we need the `s3_contents` source to remember everything it has ever seen? So that if, for example, a key is retracted from `s3_keys`, we could retract everything we downloaded for that key...

Yeah, something like this will be necessary when we want to support deletes/updates. I think that we can do this well with a normalized pair of metadata tables.

I was imagining a larger refactor (or really creation) of metadata tables for sources. Right now for [each of our input sources we insert one column of metadata][metadata], that is differently named for each source type.

I'm going to write this up as a separate issue, but I am thinking that we should instead insert *two* columns -- both of which are always the same name -- something like `mz_record_no` and `mz_metadata_id` and then create a metadata table per source-type (as well as a mapping table), so that we end up with, for S3:

```text
 +---------------|--------------|-----------------+
 | <source_cols> | mz_record_no | mz_metadata_id  |
 |               | 1            | 1               |
 |               | 2            | 1               +-
 |               | 3            | 1               | \--
 |               | 4            | 2               |    \--
 +---------------|--------------|-----------------+       \--
                                                              \--
                                                                 \--
                                                      +-------------\+----------------+
                                                      | source_table | metadata_table |
                                                      | public.ex    | materialize.ex |
 +----+--------+--------+-----------------+           +--------------+----------------+
 | id | bucket | object | object_complete |                ------/
 | 1  | ex     | a.csv  | true            |          -----/
 | 2  | ex     | b.csv  | false           |    -----/
 |    |        |        |                 |<--/
 |    |        |        |                 |
 +----+--------+--------+-----------------+
```

[metadata]: https://github.com/MaterializeInc/materialize/blob/192738f0f4ea9c16614940a77517e4f415230f82/src/dataflow-types/src/types.rs#L540-L558

**Comment by umanwizard** [comment]
> I don't think we should try to make timestamps align for two reasons:

Sure, but then how _will_ you determine what timestamp to read `s3_keys` at, for a particular timestamp of `s3_sources` ?

**Comment by quodlibetor** [comment]
Oh, I was imagining that `s3_keys` would be required to be a materialized view and we would do the equivalent of `TAIL WITH SNAPSHOT` on it.

**Comment by mjibson** [comment]
The timestamp generation here is not different from how non-BYO sources work today right? If that's the case, then I think we don't need to worry too much about transactionality guarantees because we already don't provide those. If a user needs more guarantees they would have to either use BYO or some external consistency data. We could consider supporting something (if needed by users) where they say "hey please now ingest file X for source A and file Y for source B" and we guarantee that those are provided at the same time (this is the external consistency data). But doing that with big file blobs over multiple sources seems unlikely to be a thing people will want?

**Comment by benesch** [comment]
I agree that lining the timestamps up seems messy. But I do think we should promise that if you add an object to `s3_contents` at time _t_, then data from that object won't show up before time _t_. I think this basically comes for free with wall clock time.

I also think that we should just bail on supporting BYO S3 sources for ... at least a while. I agree with Matt that the way to make BYO sources work with S3 is to disallow using the `s3_contents` view directly, and instead require that you bundle both _what objects to read_ and _what time to assign to the data within_ in the BYO stream.

The only thing that seems potentially problematic to me is deleting the records for an object if that object is retracted from `s3_contents`. Even in a world with the metadata tables, we're not going to be able to issue retractions without materializing all of `s3_contents`, right? Otherwise you won't know the full set of columns to retract.

**Comment by quodlibetor** [comment]
> But doing that with big file blobs over multiple sources seems unlikely to be a thing people will want?

Yeah I can imagine wanting it but not without serious contortions, and no one has expressed any interest in this yet. Possibly as we build out support for reading new objects folks will want something like that? It seems likely to require a fairly sophisticated Materialize user who'd be willing to work with us to make sure it works well for them.

**Comment by quodlibetor** [comment]
> Even in a world with the metadata tables, we're not going to be able to issue retractions without materializing all of s3_contents, right?

Ah, yeah we'd need to track external metadata (plausibly just some internal IDs) in some indexes somewhere across all downstream dataflows to be able to do this without materializing the sources.

But yeah we have no plans for supporting deletes or updates from unmaterialized S3 data. Even deleting from materialized S3 sources isn't on any kind of roadmap, but that one seems much more possible with reasonable effort.

**Comment by umanwizard** [comment]
A few questions:

(1) Do we have a way now to know for sure that a relation is append-only ? If so, we should just error to the psql client if the user tries to create an s3 source with object names drawn from a non-append-only source.

(2) Why was the idea of "make the source have the same timestamps as the underlying object name view" rejected? Then at least we are not introducing another source of indeterminacy into the system.

**Comment by quodlibetor** [comment]
> Do we have a way now to know for sure that a relation is append-only? If so, we should just error to the psql client if the user tries to create an s3 source with object names drawn from a non-append-only source.

I'm not sure if it's currently easy to determine this. I think it's a reasonable thing to consider as a separate issue.

> Why was the idea of "make the source have the same timestamps as the underlying object name view" rejected? Then at least we are not introducing another source of indeterminacy into the system.

I think several reasons, all stemming from the fact that these are IO-heavy/not pure functions:

1. Logically but not practically, this is inherently indeterminate, since we're loading the data and using it. Putting the data into a dataflow doesn't make it more or less deterministic than just performing a key scan in-memory and using the results.
2. If we are putting items into dataflow we can't use the results from dataflow until a capability has been closed and the timestamp has advanced. So we can't insert something into dataflow, use the results of it, and load objects based on it, in the same timestamp. You can sort of do that in an operator, but then I think that the operator is going to hold up capabilities for the entire system while you're downloading an object
3. The objects that we're downloading are almost unboundedly large (up to 5TB) if we want to hold on to a capability while we're download an object of that size then it is impossible for materialized to retire any data until we've closed out the entire object.

   1. Listing object keys can reveal up to 1,000 objects per request (although you can limit it to less than that if you'd prefer), each of which needs to be independently downloaded -- for maximum throughput you often need to download multiple objects in parallel. This amplifies the previous point.

**Comment by umanwizard** [comment]
> The objects that we're downloading are almost unboundedly large (up to 5TB)

got it, makes sense.

**Comment by ajwerner** [comment]
I'm not sure if this is the right issue to post to -- happy to split out into another one if not. CockroachDB CDC support writing to S3 using newline delimited JSON objects, as you might want for Materialize. It writes the objects in an envelope that has metadata about the stream:

```
{"table": "<table name>, "key": [<list of datums forming the primary key>], "value": {"before": {"<col1>": ...}, "after": {"col1": ...}, "updated": <hlc timestamp as decimal string>}}
```

It'd be cool if there were a way to pick apart the envelope to tell materialize about it so that it could cogently support deletes an updates powered off of these streams. Another important point about these s3 streams from cockroach is that they are intended to be consumed in lexicographical order. They expose a `RESOLVED` file which carries a timestamp that indicates that once all files up to the `RESOLVED` file have been consumed, then no new, unseen lines will be emitted. 

For what it's worth, the `"before"` portion of that messages was designed in partnership with @benesch (https://github.com/cockroachdb/cockroach/issues/28666 / https://github.com/cockroachdb/cockroach/pull/41793).

In practice, I've found these cloud storage sinks rather pleasant and operationally low-overhead relative to getting the kafka stuff all hooked up. Also, that same cloud-storage sink in cockroach will happily hit an http server that just writes to the local disk which honestly seems like a very nice way to mess around with materialize getting driven from OLTP traffic.

**Comment by umanwizard** [comment]
@ajwerner You could probably pick apart the `before`/`after` stuff and get Materialize to see it as morally inserts/updates/deletes now, in SQL, by using the `repeat` function on a diff column, e.g.

```
CREATE SOURCE s AS  ...;
CREATE VIEW befores (data, diff) AS SELECT s->before, -1 FROM s WHERE s->before IS NOT NULL;
CREATE VIEW afters (data, diff) AS SELECT s->after, 1 FROM s WHERE s->after IS NOT NULL;
CREATE VIEW all_diffs AS SELECT * FROM befores UNION ALL SELECT * FROM afters
CREATE VIEW repeated AS SELECT data FROM all_diffs, repeat(all_diffs.diff);
```

(Code untested, but something like that should work).

What's not immediately possible is ingesting the progress messages (the `RESOLVED` field) and respecting the timestamp boundaries - we have our own format for this that looks a bit difference. But it would be certainly possible in principle to support - I suggest filing another issue with a title like "support CockroachDB CDC format", for sure.

**Comment by awang** [comment]
@quodlibetor is this task done, or should we leave it open?

**Comment by quodlibetor** [comment]
@awang This is really about specifying an allowlist of objects to ingest, which we don't yet support. I updated the title to use standard jargon for that.

But no, we can't close it. I don't think it's a milestone blocker though since we've only received mild requests to implement it.

**Comment by quodlibetor** [comment]
Actually, I take that back, we should create a separate issue for the manifest, and close this one, since we do in fact have *a* way to specify the objects to ingest, it's just not fully general.
